File: leetcode/java/find1stmissingpositive.java
/*
 * Given an unsorted integer array, find the 1st missing positive integer
 *
 * test case:
 * [1,2,0], return 3
 * [4,2,3], return 1
 * [1,2,3], return 4
 * [4,1,4,3], return 2
 * */
import java.io.*;
import java.util.*;

public class find1stmissingpositive{
    public int findMissingPositive(int[] A){
        int n = A.length;
        boolean[] B = new boolean[n];   //to put [1,n]
        for(int i = 0; i < n; ++i){
            if(A[i] < 1 || A[i] > n)    continue;
            B[A[i] - 1] = true;
        }
        int j = 0;
        for(; j < n && B[j]; ++j);
        return j + 1;
    }
}

File: leetcode/src/find1stmissingpositive.cpp
/*
 * given an unsorted integer array, find the first missing positive integer
 *
 * test case:
 * [1,2,0], return 3
 * [3,4,-1,1], return 2
 * */
#include "../include/preliminary.h"

class Solution {

public:

    int findMissingPositive(int A[], int n){
        int B[n];
        memset(B, 0, sizeof(B));

        for(int i = 0; i < n; ++i){
            if(A[i] > n || A[i] < 1)    continue;
            B[A[i] - 1] = 1;
        }

        int t = 0;
        for(; t < n && B[t] != 0; ++t);

        return t+1;  // if t == n, return n+1
    }
};
